### PR TITLE
Cancel Defunct Hearing Tasks during Pending-Appeal Substitution

### PR DESCRIPTION
**Pull request by jcq**
Connects [CASEFLOW-1695](https://vajira.max.gov/browse/CASEFLOW-1695)

### Description
When `ScheduleHearingTask` is selected to be reopened during pending-appeal substitution, we need to cancel any still-open, related hearing tasks (as new ones will be created through normal processes).

These include:
* `AssignHearingDispositionTask`
* `ChangeHearingDispositionTask`
* `EvidenceSubmissionWindowTask`
* `TranscriptionTask`

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] Any open tasks of types noted above that were open prior to substitution will be cancelled after, with a `cancellation_reason` of "substitution"

### Testing Plan
1. Go to ...

### CODE REVIEW

File: spec/workflows/same_appeal_substitution_tasks_factory_spec.rb [hunk 1]
```diff
@@ -237,6 +237,41 @@
...
+          let(:schedule_hearing_task) { appeal.tasks.find_by(type: "ScheduleHearingTask") }
+          let(:assign_hearing_disposition_task) { create(:assign_hearing_disposition_task, parent: hearing_task) }
+          let!(:transcription_task) { create(:transcription_task, parent: assign_hearing_disposition_task) }
+          let!(:evidence_submission_window_task) do
+            create(:evidence_submission_window_task, parent: assign_hearing_disposition_task)
+          end
+
+          let(:selected_task_ids) { [schedule_hearing_task.id] }
+
+          let(:selected_task_ids) { [schedule_hearing_task.id] }
```
Comment (n1zyy) [inline spec/workflows/same_appeal_substitution_tasks_factory_spec.rb:253]: Is this accidental duplication?

File: app/models/task.rb [hunk 2]
```diff
@@ -55,7 +55,8 @@ class Task < CaseflowRecord
   }
 
   enum cancellation_reason: {
-    Constants.TASK_CANCELLATION_REASONS.poa_change.to_sym => Constants.TASK_CANCELLATION_REASONS.poa_change
+    Constants.TASK_CANCELLATION_REASONS.poa_change.to_sym => Constants.TASK_CANCELLATION_REASONS.poa_change,
+    Constants.TASK_CANCELLATION_REASONS.substitution.to_sym => Constants.TASK_CANCELLATION_REASONS.substitution
```
Comment (jcq) [inline app/models/task.rb:59]: Trying to match what @besampson0 added in his [related PR](https://github.com/department-of-veterans-affairs/caseflow/pull/17034)

File: app/workflows/same_appeal_substitution_tasks_factory.rb [hunk 3]
```diff
@@ -81,4 +84,23 @@ def reopen_decision_tasks
...
+      TranscriptionTask.name
+    ]
+    tasks_to_cancel = @appeal.tasks.select { |task| types_to_cancel.include?(task.type) && task.open? }
+    tasks_to_cancel.each do |task|
+      task.update!(
+        status: Constants.TASK_STATUSES.cancelled,
+        cancellation_reason: Constants.TASK_CANCELLATION_REASONS.substitution
+      )
+    end
+  end
```
Comment (jcq) [inline app/workflows/same_appeal_substitution_tasks_factory.rb:105]: Code Climate is flagging this method with this:
```
SameAppealSubstitutionTasksFactory#cancel_defunct_hearing_tasks refers to 'task' more than self (maybe move it to another class?)
```

I suppose we could create some sort of `cancel_due_to_substitution` method on `Task`, but that also doesn't seem great...?

File: client/constants/TASK_CANCELLATION_REASONS.json [hunk 4]
```diff
@@ -1,3 +1,4 @@
 {
-  "poa_change": "poa_change"
+  "poa_change": "poa_change",
+  "substitution": "substitution"
```
Comment (jcq) [inline client/constants/TASK_CANCELLATION_REASONS.json:3]: Also mimicking what was done in #17034

File: spec/workflows/same_appeal_substitution_tasks_factory_spec.rb (hunk 1 above)
Comment (jcq) [reply spec/workflows/same_appeal_substitution_tasks_factory_spec.rb:253]: Yes, yes it is. 😅

File: app/workflows/same_appeal_substitution_tasks_factory.rb [hunk 5]
```diff
@@ -81,4 +84,21 @@ def reopen_decision_tasks
...
+
+  # Called if a `ScheduleHearingTask` is selected to be reopened
+  def cancel_defunct_hearing_tasks
+    types_to_cancel = [
+      AssignHearingDispositionTask.name,
+      ChangeHearingDispositionTask.name,
+      EvidenceSubmissionWindowTask.name,
+      TranscriptionTask.name
+    ]
+    tasks_to_cancel = @appeal.tasks.select { |task| types_to_cancel.include?(task.type) && task.open? }
```
Comment (pbradin) [inline app/workflows/same_appeal_substitution_tasks_factory.rb:96]: Bryce's PR might be refactoring this a bit because of the CircleCi flag.

**Review by pbradin — APPROVED** [review]
These work fine.  I think your PR and Bryce's probably needed to be synced again before merging because of potential conflicts in the same_appeal_substitution_tasks_factory_spec.

File: app/workflows/same_appeal_substitution_tasks_factory.rb (hunk 5 above)
Comment (n1zyy) [reply app/workflows/same_appeal_substitution_tasks_factory.rb:97]: There's some [Slack discussion](https://dsva.slack.com/archives/CQCUMFA0H/p1639602432210200?thread_ts=1639601434.209900&cid=CQCUMFA0H) about this; this is throwing a FeatureEnvy warning. Our [style guide](https://github.com/department-of-veterans-affairs/caseflow/wiki/Ruby-Style-Guide#reekfeatureenvy) touches on this a bit.

Taking its suggestion at face value, we refer to Task a lot so perhaps this logic might make more sense in task.rb. But, as we discussed, it seems like the answer to that question is "No." You could move various parts of this into Task, but they really don't make a whole lot of sense there.

+1 to skipping this alert IMHO.

**Pull request merged**